### PR TITLE
Lower ResizeNearestNeighbor when half_pixel_centers=true

### PR DESCRIPTION
**Pull request by WindQAQ**
Lower ResizeNearestNeighbor when `half_pixel_centers=true`. This PR makes the lowering pattern match the use in TF python library as combination of `align_corners=false` and `half_pixel_centers=true` is the only code path in `tf.image.resize`. Test thoroughly in https://colab.research.google.com/drive/1YCaLPOaagFzuDRg4M67bTUz1boatCvhf?usp=sharing.

Also remove some unnecessary reshape
- Before AddV2: x_indices can be broadcasted
- Before and after GatherV2: indices can be 2D without back and forth reshape

### CODE REVIEW

**Comment by gbaned** [comment]
@WindQAQ  Can you please resolve conflicts? Thanks!

**Comment by gbaned** [comment]
@WindQAQ Can you please resolve conflicts? Thanks!

**Comment by gbaned** [comment]
@WindQAQ Any update on this PR? Please. Thanks!

File: tensorflow/compiler/mlir/tensorflow/tests/lower_tf.mlir [hunk 1]
```diff
@@ -830,112 +830,183 @@ func @imag_resize_nearest(%arg0: tensor<1x7x7x1xi32>) -> tensor<1x3x3x1xi32> {
   %shape = "tf.Const"() {device = "", value = dense<3> : tensor<2xi32>} : () -> tensor<2xi32>
 
   // CHECK: [[VAL0:%.+]] = "tf.Const"() {value = dense<1> : tensor<i32>}
-  // CHECK: [[VAL1:%.+]] = "tf.Const"() {value = dense<[1, 3, 3, 1]>
```
Comment (tenjua) [inline tensorflow/compiler/mlir/tensorflow/tests/lower_tf.mlir:833]: Patch O. diff

File: tensorflow/compiler/mlir/tensorflow/transforms/lower_tf.cc [hunk 2]
```diff
@@ -1493,13 +1470,10 @@ class LowerResizeNearestNeighbor : public RewritePattern {
 
     Value gathered_values = rewriter.create<GatherV2Op>(
         loc,
-        RankedTensorType::get({batch_cst, out_spatial_cst, channels_cst},
+        RankedTensorType::get({batch_cst, out_y_cst, out_x_cst, channels_cst},
                               input_element_ty),
         input_collapsed_spatial, indices, /*axis=*/one);
 
-    gathered_values =
```
Comment (tenjua) [inline tensorflow/compiler/mlir/tensorflow/transforms/lower_tf.cc:1500]: n Y p

File: tensorflow/compiler/mlir/tensorflow/tests/lower_tf.mlir (hunk 1 above)
Comment (tenjua) [reply tensorflow/compiler/mlir/tensorflow/tests/lower_tf.mlir:833]: . Patch

File: tensorflow/compiler/mlir/tensorflow/transforms/lower_tf.cc [hunk 3]
```diff
@@ -1390,9 +1390,9 @@ class LowerResizeNearestNeighbor : public RewritePattern {
     Value zero_f32 = rewriter.create<ConstOp>(
         loc, GetScalarOfType(rewriter.getF32Type(), 0.0));
     Value one_f32 = rewriter.create<ConstOp>(
-        loc, GetScalarOfFloatType(rewriter.getF32Type(), 1.0));
```
Comment (tenjua) [inline tensorflow/compiler/mlir/tensorflow/transforms/lower_tf.cc]: N Y P

File: tensorflow/compiler/mlir/tensorflow/transforms/lower_tf.cc [hunk 4]
```diff
@@ -1493,13 +1470,10 @@ class LowerResizeNearestNeighbor : public RewritePattern {
...
+        RankedTensorType::get({batch_cst, out_y_cst, out_x_cst, channels_cst},
                               input_element_ty),
         input_collapsed_spatial, indices, /*axis=*/one);
 
-    gathered_values =
-        rewriter.create<ReshapeOp>(loc, result_ty, gathered_values, out_shape);
-
     rewriter.replaceOp(op, gathered_values);
     return success();
   }
```
Comment (tenjua) [inline tensorflow/compiler/mlir/tensorflow/transforms/lower_tf.cc:1479]: N Y P

**Review by tenjua — APPROVED** [review]

File: tensorflow/compiler/mlir/tensorflow/tests/lower_tf.mlir [hunk 5]
```diff
@@ -830,112 +830,183 @@ func @imag_resize_nearest(%arg0: tensor<1x7x7x1xi32>) -> tensor<1x3x3x1xi32> {
...
+  return %resize: tensor<1x3x3x1xi32>
+}
+
 // CHECK-LABEL: func @imag_resize_nearest_dyn_img
 func @imag_resize_nearest_dyn_img(%arg0: tensor<1x?x?x1xi32>) -> tensor<1x3x3x1xi32> {
   %shape = "tf.Const"() {device = "", value = dense<3> : tensor<2xi32>} : () -> tensor<2xi32>
 
   // CHECK: [[VAL0:%.+]] = "tf.Const"() {value = dense<1> : tensor<i32>}
   // CHECK: [[VAL1:%.+]] = "tf.Const"() {value = dense<[3, 1]> : tensor<2xi32>}
-  // CHECK: [[VAL2:%.+]] = "tf.Const"() {value = dense<9> : tensor<1xi32>}
```
Comment (tenjua) [inline tensorflow/compiler/mlir/tensorflow/tests/lower_tf.mlir:850]: . diff

**Review by tenjua — CHANGES_REQUESTED** [review]

**Comment by gbaned** [comment]
@WindQAQ  Can you please check @tenjua's comments and keep us posted ? Thanks!

**Comment by gbaned** [comment]
@WindQAQ  Any update on this PR? Please. Thanks!

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by gbaned** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!

**Pull request closed without merging**